crates/starknet-devnet/tests/test_account_impersonation.rs

```diff
@@ -0,0 +1,315 @@
+
+        let invoke_calls = vec![get_invoke_transaction_request(AMOUNT_TO_TRANSFER)];
+
+        // vector of tuples of steps (impersonation action, do validation, expected_result)
+        let steps = vec![
+            (
+                Some(ImpersonationAction::ImpersonateAccount(account.address())),
+                true,
+                TestCaseResult::Success,
+            ),
+            (
+                Some(ImpersonationAction::ImpersonateAccount(account.address())),
+                false,
+                TestCaseResult::Success,
+            ),
+            (None, false, TestCaseResult::Success),
+            (None, true, TestCaseResult::Failure { msg: "invalid signature".to_string() }),
+            (Some(ImpersonationAction::AutoImpersonate), true, TestCaseResult::Success),
+            (Some(ImpersonationAction::AutoImpersonate), false, TestCaseResult::Success),
+        ];
+
+        for (impersonation_action_option, do_validate, expected_result) in steps {
+            if let Some(impersonation_action) = impersonation_action_option {
+                forked_devnet.impersonate_account(&impersonation_action).await.unwrap();
+            }
+
+            let simulation_result =
+                account.execute(invoke_calls.clone()).simulate(!do_validate, false).await;
+            match expected_result {
+                TestCaseResult::Success => {
+                    simulation_result.expect("Expected simulation to succeed");
+                }
+                TestCaseResult::Failure { msg } => {
+                    let err = simulation_result.err().unwrap();
+                    assert!(format!("{:?}", err).to_lowercase().contains(&msg));
+                }
+            }
+
+            forked_devnet.restart().await.unwrap();
+        }
+    }
+
+    async fn test_declare_transaction(
+        origin_devnet: &BackgroundDevnet,
+        impersonation_actions: &[ImpersonationAction],
+        expected_result: TestCaseResult,
+    ) {
+        let forked_devnet = origin_devnet.fork().await.unwrap();
+
+        let mut account =
+            get_account_interacting_with_forked_devnet(origin_devnet, &forked_devnet).await;
+
+        for action in impersonation_actions.iter() {
+            forked_devnet.impersonate_account(action).await.unwrap();
+        }
+
+        let sierra_path = format!("{}/{}", env!("CARGO_MANIFEST_DIR"), CAIRO_1_CONTRACT_PATH);
+        let contract_artifact: SierraClass =
+            serde_json::from_reader(std::fs::File::open(sierra_path).unwrap()).unwrap();
+
+        let casm_path = format!("{}/{}", env!("CARGO_MANIFEST_DIR"), CAIRO_1_CASM_PATH);
+        let casm_contract_definition: CompiledClass =
+            serde_json::from_reader(std::fs::File::open(casm_path).unwrap()).unwrap();
+        let compiled_class_hash = (casm_contract_definition.class_hash()).unwrap();
+        assert_eq!(Felt::from(compiled_class_hash).to_prefixed_hex_str(), CASM_COMPILED_CLASS_HASH);
```

Comment:
Can it be replaced with `get_simple_contract_in_sierra_and_compiled_class_hash`? I assume you copied it from Mikolaj's test, but perhaps this utility function could also be used there. If you want, we can leave this for a future PR.

Reply:
[5e6d460](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/436/commits/5e6d460a05e9525452001d3197d7815dea463f41)